flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java

```diff
@@ -114,6 +115,21 @@ protected void respondAsLeader(ChannelHandlerContext ctx, RoutedRequest routedRe
 			log.trace("Received request " + httpRequest.uri() + '.');
 		}
 
+		synchronized (this) {
+			if (terminationFuture != null) {
+				String errorMsg = "The handler instance for " + untypedResponseMessageHeaders.getTargetRestEndpointURL()
+					+ " had already been closed";
+				log.warn(errorMsg);
+				HandlerUtils.sendErrorResponse(
+					ctx,
+					httpRequest,
+					new ErrorResponseBody(errorMsg),
+					HttpResponseStatus.BAD_REQUEST,
+					responseHeaders);
+				return;
+			}
+		}
+
 		FileUploads uploadedFiles = null;
 		try {
 			inFlightRequestTracker.registerRequest();
```

Comment:
this also needs to happen under the lock, otherwise the race condition isn't fixed. (just move the entire synchronized block here)